### ISSUE TITLE
Win32Exception

### ISSUE THREAD

**Issue by matthewpetch** [MerrionComputing/PrintQueueWatch#2]
PrinterInformationChanged works fine but PrinterMonitor_JobAdded throws a Win32Exception (The parameter is incorrect) in SpoolerStructs.vb, line 75 and line 207

Running as either 32 or 64bit, Windows 10 64bit 

If I ignore Win32Exceptions, it then throws

System.Runtime.InteropServices.SEHException: 'External component has thrown an exception.'

In PrinterChangeNotificationThread, line 90

**Comment by MerrionComputing** [comment]
_Initial thoughts._
It looks like the first call to GetJob is failing where it tries to get the buffer size (line 70)

`If Not GetJob(hPrinter, dwJobId, 1, ptBuf, 0, BytesWritten) Then`

The API declaration is:-
```
BOOL GetJob(
  _In_  HANDLE  hPrinter,
  _In_  DWORD   JobId,
  _In_  DWORD   Level,
  _Out_ LPBYTE  pJob,
  _In_  DWORD   cbBuf,
  _Out_ LPDWORD pcbNeeded
);
```

This rules out the JOB_INFO_n structures (I think)

**Comment by matthewpetch** [comment]
Sorry, after further testing this is only an issue on 64 bit builds

I've noticed it still throws the SEHException occasionally on 32 bit, but it still manages to fire the event for an event for JobAdded

**Comment by joelharkes** [comment]
I seem on have the same problem. At least in 1 c# project the job events don't fire becuase it seems unable to read the right job id (it always reads 0). in sturct: PrinterNotifyInfoData property: dwId.

**Comment by joelharkes** [comment]
Seem to get the same symptoms where job events don't fire only printer information events.

seems to occur in the following code (i migrated it to c# instead):

```
public PrinterNotifyInfo(IntPtr mhPrinter, IntPtr lpAddress, ref PrintJobCollection PrintJobs)
    {
      if (PrinterNotifyInfoData.TraceSwitch.TraceVerbose)
        Trace.WriteLine("New(" + mhPrinter.ToString() + "," + lpAddress.ToString() + ")", this.GetType().ToString());

      if (!(lpAddress.ToInt64() == 0))
      {
        // \\ Create the array list of jobs involved in this event
        colPrintJobs = new ArrayList();

        // \\ Read the data of this printer notification event
        Marshal.PtrToStructure(lpAddress, msInfo);

        int nInfoDataItem;
        // \\ Offset the pointer by the size of this class
        var lOffset = lpAddress + Marshal.SizeOf(msInfo);

        // \\ Process the .adata array
        for (nInfoDataItem = 0; nInfoDataItem <= msInfo.Count - 1; nInfoDataItem++)
        {
          var itemdata = new PrinterNotifyInfoData(lOffset);
          if (itemdata.Type == Printer_Notification_Types.JOB_NOTIFY_TYPE)
          {
            if (itemdata.dwId == 0)
            {
              if (PrinterMonitorComponent.ComponentTraceSwitch.TraceWarning)
                Trace.WriteLine("JOB_NOTIFY_TYPE has zero job id ");
```

itemdata.dwId  is zero and that's why it doesn't fire any job events somehow. I indeed suspect it has to do with data allignment, but this is not my expertise. I will let you know if i find anything. Any help would be appreciated too.

**Comment by joelharkes** [comment]
Removing The reserved property makes it work buut i think (im pretty sure) i'll have to find a better way to fix this:
```
[StructLayout(LayoutKind.Sequential)]
  class PrinterNotifyInfoData
  {
    /// <summary>
    /// Indicates the type of information provided. This member can be one of the following values.
    /// JOB_NOTIFY_TYPE
    /// PRINTER_NOTIFY_TYPE
    /// </summary>
    public short wType;
    /// <summary>
    /// Indicates the field that changed. For a list of possible values, see the Remarks section.
    /// </summary>
    public short wField;
    //public int dwReserved;
    public int dwId;
    public int cbBuff;
    public IntPtr pBuff;
```

**Comment by MerrionComputing** [comment]
The class should match [PRINTER_NOTIFY_INF_DATA](https://docs.microsoft.com/en-us/windows/desktop/printdocs/printer-notify-info-data)

```
typedef struct _PRINTER_NOTIFY_INFO_DATA {
  WORD  Type;
  WORD  Field;
  DWORD Reserved;
  DWORD Id;
  union {
    DWORD  adwData[2];
    struct {
      DWORD  cbBuf;
      LPVOID pBuf;
    } Data;
  } NotifyData;
}
```

I think explicit use of Int32 for DWORD and Int16 for SHORT might be better (?)

**Comment by joelharkes** [comment]
Ugh wow thanks this seems to work apparently going to short changes the data alignment? thanks resharper/vs for suggesting auto change -.- 😢

**Comment by joelharkes** [comment]
Doesnt really seem to matter, only way to fix it is to run in 32bit mode, than it starts working instantly again.

how does this data alignment work, how can i make this workin 64bit c#?

**Comment by joelharkes** [comment]
Found a solution for this problem on an old code croject reaction:

all i got it working on a 64 bit machine just do this 

Dim lOffset As Integer = mlpPrinter.ToInt32 + Marshal.SizeOf(msInfo) + 4

this will do the trick as 64 does not start on zero but increments at 8 bytes

I guess I need to fill up PRINTER_NOTIFY_INFO struct.

i hope this will be enough to make it all work

**Comment by matthewpetch** [comment]
If it works please make a PR @joelharkes

**Comment by joelharkes** [comment]
Leads to more problems throwing exceptions trying to:

```
                mlPrinterInfoChanged = System.Convert.ToInt32(mlPrinterInfoChanged + (Math.Pow(2, (int)itemdata.Field)));
```

Seems Field get unexpected large number.. 
Other people say to change some integers into int pointers, but didnt have time to test this yet..

See here: https://www.codeproject.com/articles/3313/webcontrols/?fid=13704&df=90&mpp=25&prof=True&sort=Position&view=Normal&spc=Relaxed&fr=91

If you need the project to be compatible with both architectures, you should replace some types into IntPtr

PRINTER_NOTIFY_INFO.Count
PRINTER_NOTIFY_INFO_DATA.Id
PRINTER_NOTIFY_INFO_DATA.adwData
PRINTER_NOTIFY_INFO_DATA.cbBuf
PRINTER_NOTIFY_INFO_DATA.pBuf

havent tried this yet.

**Comment by MerrionComputing** [comment]
I would think that any pointers would need no be converted, but no reason why Id and Count would?
As you say the BYTE boundary alignment may also need to be taken into account...

**Comment by joelharkes** [comment]
He is right:

PRINTER_NOTIFY_INFO.Count
P => fills up the gap 


PRINTER_NOTIFY_INFO_DATA.cbBuf = IntPtr acording to docs: (LPVOID): https://docs.microsoft.com/en-us/windows/desktop/printdocs/printer-notify-info-data
But it should fit together with: pBuf in an DWORD[2] Array, that makes no sense: 12 bytes in 8 bytes :o 
So both should actually be IntPtrs and it is fixed :).

**Comment by joelharkes** [comment]
OR add + 4 add beginning 

and + 8 for every item in the loop. (gives better result somehow).

this still doesnt completly fix it though.

This one seems to perform better, i'd advise to use this.

**Comment by joelharkes** [comment]
now getjob doesnt work yet either, working on that one now (says paramter is incorrect on calling the printerspooler.getjob()

**Comment by joelharkes** [comment]
_Out_ LPDWORD pcbNeeded seems to be a Int32 instead of a IntPrt..

**Comment by MerrionComputing** [comment]
As "pack" is not explicitly specified for within the structure it should be OK just using the [StructlayoutAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.structlayoutattribute.pack?redirectedfrom=MSDN&view=netframework-4.7.2).
The problem occurs when we get back an array as we need to explicitly jump over any unused space at the end of the structure to get the start of the next one.
Like in this [stackoverflow](https://stackoverflow.com/questions/21687648/marshalling-struct-with-array-of-struct-member)

**Comment by joelharkes** [comment]
Ok i said more than actually was the problem.

how i fixed it:
PrinterNotifyInfoData:

```
    public IntPtr cbBuff;
    public IntPtr pBuff;

  public int ToInt32()
    {
      return cbBuff.ToInt32();
    }

// in constructor:

var lOffset = lpAddress
          + Marshal.SizeOf(msInfo)
          + (Platform.Is64Bit() ? 4 : 0); //data-alignment: need to fill up the 3 integers in 64 bit.

```

The other issues i had with printjobs was with debugging: the job was already spooled in debugging and thus the job info could no longer be found.

**Comment by gustavofbreunig** [comment]
@joelharkes yes, I believe it fixes the problem here too, can you make a pull request? @MerrionComputing can you test it?

**Comment by MerrionComputing** [comment]
Yes indeed - @joelharkes if you make a pull request I will test it and merge.

**Comment by joelharkes** [comment]
Tried what i could: https://github.com/MerrionComputing/PrintQueueWatch/pull/5

#5 

not sure if this is actually correct VB (i fixed it in my repo in c#)
